### ISSUE TITLE
Kitchen fails on multiple reboots of node during run. 

### ISSUE THREAD

**Issue by sflaherty2009** [test-kitchen/test-kitchen#1376]
## Description

When running test kitchen and setting the following reboot settings. 

```
    name: chef_zero
    retry_on_exit_code:
      - 35 # 35 is the exit code signaling that the node is rebooting
      - 1
    max_retries: 3
    multiple_converge: 3
    wait_for_retry: 120
    client_rb:
      exit_status: :enabled # Opt-in to the standardized exit codes
      client_fork: false  # Forked instances don't return the real exit code
```

See kitchen.yml as follows. 
```
---
  driver:
    name: azurerm
    
  driver_config:
    subscription_id: 
    location: East US 2
    machine_size: Standard_D1
  
  transport:
    ssh_key: ~/.ssh/id_rsa
  
  provisioner:
    name: chef_zero
    retry_on_exit_code:
      - 35 # 35 is the exit code signaling that the node is rebooting
      - 1
    max_retries: 3
    multiple_converge: 3
    wait_for_retry: 120
    client_rb:
      exit_status: :enabled # Opt-in to the standardized exit codes
      client_fork: false  # Forked instances don't return the real exit code
  
  platforms:
    - name: centos-7.4
      driver_config:
        image_urn: OpenLogic:CentOS:7.4:latest
        # vnet and subnet id will work in web technology prod only. Disable otherwise.
        vnet_id: /subscriptions/9fbf7025-df40-4908-b7fb-a3a2144cee91/resourceGroups/AZ-RG-Network/providers/Microsoft.Network/virtualNetworks/AZ-VN-EastUS2-02
        subnet_id: AZ-SN-dvo
        vm_name: testEnvSolr
        vm_tags:
          ostype: linux
          distro: centos
  
  suites:
    - name: default
      data_bags_path: "test/integration/data_bags" 
      run_list:
        - recipe[cb_env_solr::default]
      attributes:
  
```
kitchen converge survives only one restart of the machine. Upon the second reboot kitchen fails. 

## Kitchen Version

Test Kitchen version 1.16.0

```ruby
 Recipe: cb_dvo_resolveDNS::linux
         * reboot[Reboot VM] action reboot_now[2018-03-01T17:20:39+00:00] WARN: Rebooting system immediately, requested by 'Reboot VM'
       
           
           Running handlers:
           Running handlers complete
           Chef Client finished, 39/52 resources updated in 02 minutes 38 seconds
       [2018-03-01T17:20:39+00:00] WARN: Rebooting server at a recipe's request. Details: {:delay_mins=>0, :reason=>"Must restart to enable network changes.", :timestamp=>2018-03-01 17:20:39 +0000, :requested_by=>"Reboot VM"}
           
           Running handlers:
       [2018-03-01T17:20:39+00:00] ERROR: Running exception handlers
           Running handlers complete
       [2018-03-01T17:20:39+00:00] ERROR: Exception handlers complete
           Chef Client failed. 39 resources updated in 02 minutes 38 seconds
```

## ChefDK Version

```ruby
Chef Development Kit Version: 1.3.40
chef-client version: 12.19.36
delivery version: master (69bfa4a76959a8d093511a90fddd7a1f7e43e354)
berks version: 5.6.4
kitchen version: 1.16.0
```

## Ruby Version

If you are not using test-kitchen via ChefDK, please provide the output of ruby --version.

```ruby
ruby 2.0.0p648 (2015-12-16) [x86_64-linux]
```

## Platform Version

Linux CentOS 7.4

## Replication Case

Multiple runs of the following ruby block. 

ruby 
```
reboot 'Reboot VM' do
  action :reboot_now
  reason 'Must restart to enable network changes.'
  notifies :run, 'execute[Run chef-client after restart]', :before
  only_if { node.run_state['reboot'] == 'true' }
end
``` 

for information on how to create a good replication case.

## Kitchen Output

[output.txt](https://github.com/test-kitchen/test-kitchen/files/1772732/output.txt)

## Kitchen Diagnose

Please include the output of `kitchen diagnose --all` or `kitchen diagnose --loader` (if the first failed).  This can be helpful in troubleshooting, so please include it in a gist.

ruby 
```---
timestamp: 2018-03-01 21:08:16 UTC
kitchen_version: 1.16.0
plugins:
  driver:
    Azurerm:
      class: Kitchen::Driver::Azurerm
      version: 
      api_version: 
  provisioner:
    ChefZero:
      class: Kitchen::Provisioner::ChefZero
      version: 1.16.0
      api_version: 2
  transport:
    Ssh:
      class: Kitchen::Transport::Ssh
      version: 1.16.0
      api_version: 1
  verifier:
    Busser:
      class: Kitchen::Verifier::Busser
      version: 1.16.0
      api_version: 1
loader:
  process_erb: true
  process_local: true
  process_global: true
  global_config: 
  project_config:
    filename: "/data/cookbooks/cb_env_solr/.kitchen.yml"
    raw_data:
      driver:
        name: azurerm
      driver_config:
        subscription_id: 
        location: East US 2
        machine_size: Standard_D1
      transport:
        ssh_key: "~/.ssh/id_rsa"
      provisioner:
        name: chef_zero
        retry_on_exit_code:
        - 35
        - 1
        max_retries: 3
        multiple_converge: 3
        wait_for_retry: 120
        client_rb:
          exit_status: ":enabled"
          client_fork: false
      platforms:
      - name: centos-7.4
        driver_config:
          image_urn: OpenLogic:CentOS:7.4:latest
          vnet_id: "/subscriptions/9fbf7025-df40-4908-b7fb-a3a2144cee91/resourceGroups/AZ-RG-Network/providers/Microsoft.Network/virtualNetworks/AZ-VN-EastUS2-02"
          subnet_id: AZ-SN-dvo
          vm_name: testEnvSolr
          vm_tags:
            ostype: linux
            distro: centos
      suites:
      - name: default
        data_bags_path: test/integration/data_bags
        run_list:
        - recipe[cb_env_solr::default]
        attributes: 
  local_config: 
  combined_config:
    filename: 
    raw_data:
      driver:
        name: azurerm
      driver_config:
        subscription_id: 
        location: East US 2
        machine_size: Standard_D1
      transport:
        ssh_key: "~/.ssh/id_rsa"
      provisioner:
        name: chef_zero
        retry_on_exit_code:
        - 35
        - 1
        max_retries: 3
        multiple_converge: 3
        wait_for_retry: 120
        client_rb:
          exit_status: ":enabled"
          client_fork: false
      platforms:
      - name: centos-7.4
        driver_config:
          image_urn: OpenLogic:CentOS:7.4:latest
          vnet_id: "/subscriptions/9fbf7025-df40-4908-b7fb-a3a2144cee91/resourceGroups/AZ-RG-Network/providers/Microsoft.Network/virtualNetworks/AZ-VN-EastUS2-02"
          subnet_id: AZ-SN-dvo
          vm_name: testEnvSolr
          vm_tags:
            ostype: linux
            distro: centos
      suites:
      - name: default
        data_bags_path: test/integration/data_bags
        run_list:
        - recipe[cb_env_solr::default]
        attributes: 
instances:
  default-centos-74:
    platform:
      os_type: unix
      shell_type: bourne
    state_file: {}
    driver:
      azure_environment: Azure
      azure_resource_group_name: default-centos-74
      azure_resource_group_prefix: kitchen-
      azure_resource_group_suffix: ''
      boot_diagnostics_enabled: 'true'
      custom_data: ''
      data_disks: 
      existing_storage_account_blob_url: ''
      existing_storage_account_container: vhds
      explicit_resource_group_name: 
      format_data_disks: false
      format_data_disks_powershell_script: false
      image_id: ''
      image_url: ''
      image_urn: OpenLogic:CentOS:7.4:latest
      kitchen_root: "/data/cookbooks/cb_env_solr"
      location: East US 2
      log_level: :info
      machine_size: Standard_D1
      name: azurerm
      os_disk_size_gb: ''
      os_type: linux
      password: P2ssw0rd
      pre_deployment_parameters: {}
      pre_deployment_template: ''
      public_ip: false
      storage_account_type: Standard_LRS
      subnet_id: 
      subscription_id: 
      test_base_path: "/data/cookbooks/cb_env_solr/test/integration"
      use_managed_disks: true
      username: azure
      vm_name: testEnvSolr
      vm_tags:
        ostype: linux
        distro: centos
      vnet_id: "/subscriptions/9fbf7025-df40-4908-b7fb-a3a2144cee91/resourceGroups/AZ-RG-Network/providers/Microsoft.Network/virtualNetworks/AZ-VN-EastUS2-02"
      winrm_powershell_script: false
    provisioner:
      always_update_cookbooks: false
      attributes: {}
      chef_client_path: "/bin/chef-client"
      chef_omnibus_install_options: 
      chef_omnibus_url: https://omnitruck.chef.io/install.sh
      chef_zero_host: 
      chef_zero_port: 8889
      client_rb:
        exit_status: ":enabled"
        client_fork: false
      clients_path: 
      command_prefix: 
      config_path: 
      cookbook_files_glob: README.*,metadata.{json,rb},attributes/**/*,definitions/**/*,files/**/*,libraries/**/*,providers/**/*,recipes/**/*,resources/**/*,templates/**/*
      data_bags_path: "/data/cookbooks/cb_env_solr/test/integration/data_bags"
      data_path: 
      deprecations_as_errors: false
      encrypted_data_bag_secret_key_path: 
      enforce_idempotency: false
      environments_path: 
      ftp_proxy: 
      http_proxy: 
      https_proxy: 
      json_attributes: true
      kitchen_root: "/data/cookbooks/cb_env_solr"
      log_file: 
      log_level: auto
      max_retries: 3
      multiple_converge: 3
      name: chef_zero
      named_run_list: {}
      nodes_path: 
      policyfile: 
      policyfile_path: 
      profile_ruby: false
      require_chef_omnibus: true
      retry_on_exit_code:
      - 35
      - 1
      roles_path: 
      root_path: "/tmp/kitchen"
      ruby_bindir: "/embedded/bin"
      run_list:
      - recipe[cb_env_solr::default]
      sudo: true
      sudo_command: sudo -E
      test_base_path: "/data/cookbooks/cb_env_solr/test/integration"
      wait_for_retry: 120
    transport:
      compression: false
      compression_level: 0
      connection_retries: 5
      connection_retry_sleep: 1
      connection_timeout: 15
      keepalive: true
      keepalive_interval: 60
      kitchen_root: "/data/cookbooks/cb_env_solr"
      log_level: :info
      max_ssh_sessions: 9
      max_wait_until_ready: 600
      name: ssh
      port: 22
      ssh_gateway: 
      ssh_gateway_username: 
      ssh_key: "/root/.ssh/id_rsa"
      test_base_path: "/data/cookbooks/cb_env_solr/test/integration"
      username: root
    verifier:
      busser_bin: "/tmp/verifier/bin/busser"
      chef_omnibus_root: "/opt/chef"
      command_prefix: 
      ftp_proxy: 
      http_proxy: 
      https_proxy: 
      kitchen_root: "/data/cookbooks/cb_env_solr"
      log_level: :info
      name: busser
      root_path: "/tmp/verifier"
      ruby_bindir: "/opt/chef/embedded/bin"
      sudo: true
      sudo_command: sudo -E
      suite_name: default
      test_base_path: "/data/cookbooks/cb_env_solr/test/integration"
      version: busser
```

## NOTE:

This repository is for core issues with the Test-Kitchen framework, so if the issue is with a plugin you may be referred to file the issue with the appropriate plugin.

**Comment by cheeseplus** [comment]
There are a few things going on here: old kitchen, old chef, reboot support, and multiple_converge.

- old chef means that proper forking on reboot is busted at the client level https://github.com/chef/chef/issues/5026
- old kitchen means that reboot support doesn't have some fixes it needs
- confused if you want reboots or multi_converge, these are very different things

**Comment by thewyzard44** [comment]
having recently looked at this code segment, I would also wonder if `multiple_converge: 3` combined with rebooting would result in, potentially, 2 converges, e.g., and then a reboot, and then a fresh attempt at 3 converges via https://github.com/test-kitchen/test-kitchen/blob/master/lib/kitchen/provisioner/chef_base.rb#L609-L611 as the conerge starts over?

The combinations of these options may need better specified

**Comment by thewyzard44** [comment]
ok...  I just couldn't get my head around that combination of options (late) last night, but I was intrigued, and just had to find out.  multiple_converge and reboots do play together, but it will result in a larger number of converges like I described.

Beyond that, I am able to replicate your issue.  I found that if you address the older versions of the software, that even with the newest versions, the combination of chef-client + reboot + ssh is still racy.  What's happening is that the reboot starts before the chef-client has fully exited and made its exit status available.  Once that reboot starts, if sshd gets shut down before the chef client finishes, then your ssh connection can't get the exit status and errors out with the error you described above (which was a transport failure, and not exit code 35).  If the chef-client ever wins that race and returns 35, and ssh picks that up before getting disconnected, then kitchen would retry as expected (with a whole new set of 3 converges, in this case).  I can get random successes locally, but it fails more often than not.

After some code tracing I was only able to find one workaround:  Specify `delay_mins 1` on your reboot resource so that the chef-client has time to fully exit before the reboot starts.  It'll slow down your run, but it does defeat the race condition.
